<h1> Linear Programming </h1><br>
This notebook is for linear programming using scipy and pulp. Several simple optimization problems are covered for the purpose of illustration and to learn how to use python for this task.

In [1]:
c = [-1, 4]

In [2]:
A = [[-3, 1], [1, 2]]

In [3]:
b = [6, 4]

In [7]:
x0_bounds = (None, None)

In [8]:
x1_bounds = (-3, None)
from scipy.optimize import linprog
res = linprog(c, A_ub=A, b_ub=b, bounds=(x0_bounds, x1_bounds),options={"disp": True})

Optimization terminated successfully.
         Current function value: -22.000000  
         Iterations: 1


In [9]:
print(res)

     fun: -22.0
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([ 39.,   0.])
  status: 0
 success: True
       x: array([ 10.,  -3.])


<h2>Example 1</h2><br>
Three products are processed through three different operations. The time in munutes required per unit of each product, 
the daily capacity of the operations (in minutes per day) and the profit per unit sold of each product (in pounds) are as given
in the table.
The zero times indicate that the product does not require the given operation. It is assumed that all units produced
are sold. The given profits per unit are net values that result after all pertinent expenses are deducted. Formulate a linear 
programming model that fits the daily production. <br> <br>
<table>
<tr></tr>
<tr>
    <td> Operation </td><td> Product 1 </td><td> Product 2 </td><td> Product 3 </td><td> Operation capcity (minutes/day </td>
</tr>
<tr>
    <td> 1 </td><td> 1 </td><td> 2 </td><td> 1 </td><td> 430 </td>
</tr>
<tr>
    <td> 2 </td><td> 3 </td><td> 0 </td><td> 2 </td><td> 460 </td>
</tr>
<tr>
    <td> 3 </td><td> 1 </td><td> 4 </td><td> 0 </td><td> 420 </td>
</tr>
<tr>
    <td> Profit/unit (£) </td><td> 3 </td><td> 2 </td><td> 5 </td><td>  </td>
</tr>
</table>

Solution: The main elements of a linear programming model are (1) the variables or unknowns. (2) the objective function and <br>(3) the constraints
   <br><br> The variables are identified as the daily number of units to be manufactured of each product.
   <br> $x_1, x_2, x_3$ 
   <br>shall be the number of units produced of products 1, 2 and 3. Because of the assumption that all units produced are sold, the total profit in pounds is <br>
   $x_0 = 3x_1 + 2x_2+5x_3 \quad (1)$ <br>
   The constraints of the problem must ensure that the total processing time required by all produced units does not exceed the daily capacity of each operation. <br>
   $1x_1 + 2x_2+1x_3 \leq 430 $ <br>
   $3x_1 + 0x_2+2x_3 \leq 460 $ <br>
   $1x_1 + 4x_2+0x_3 \leq 420 $ <br>
   and it is assumed non-negative <br> $x_1 \gt 0,x_2\gt 0, x_3\gt0  $ <br>
   <br> So the task is to maximize (1) subject to the constraints.

In [10]:
X0 = [-3., -2.,-5.]
A1 = [[1.,2.,1.],[3.,0,2.],[1., 4.,0]]

B1 = [430., 460., 420.]
x0_bounds = (0., None)
x1_bounds = (0., None)
x2_bounds = (0., None)
from scipy.optimize import linprog
res = linprog(X0, A_ub=A1, b_ub=B1,options={"disp": True, "maxiter":3})

Optimization terminated successfully.
         Current function value: -1350.000000
         Iterations: 2


In [11]:
print(res)

     fun: -1350.0
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([  0.,   0.,  20.])
  status: 0
 success: True
       x: array([   0.,  100.,  230.])


I will assume that this means we produce 100 of product 2 and 230 of product 3

In [12]:
100 * 2 +230*5

1350

In [13]:
timeForOp1Prod2 = 2*100

In [14]:
timeForOp3Prod1 = 4*100

In [15]:
timeForOp1Prod3 = 230

In [16]:
timeForOp2Prod3 = 2*230

In [17]:
times=[ timeForOp1Prod2+timeForOp1Prod3,timeForOp2Prod3, timeForOp3Prod1]

In [18]:
times

[430, 460, 400]

It is clear that after 2 iterations the linear solver is almost optimal, the values should be integers becase you cannot make half of a product. The people working in operation 3 get an extra tea break

<h2> Example 2 </h2>
<p> One of the most successful applications of linear programming deals with the determination of optimal feed mix for an animal at the least cost. The model assumes the availability of certain feedstuffs or ingredients from which the feed is mixed. 
The nutritive content of each ingredient is known. The constraints of the model include (1) daily nutrient requirements of the animal (2) physical or nonnutritive limitations such as supply, feed texture, or ability of feed to be pelleted. The objective is to minimize the total cost of a given batch size of the feed mix such that the nutritive and physical constraints are satisfied. <br>
<br> Assuming the required daily batch of the feed mix is 100 pounds. The diet must contain <br>
1) At least 0.8 percent but not more than 1.2 percent calcium <br>
2) At least 22 percent protein <br>
3) At most  percent crude fibre.
<br><br>
The main ingredients or feedstuffs used include limestone (calcium carbonate), corn, and soybean meal. The nutritive content of these ingredients is summarized in the table
<br><br>
<table>
<tr>
    <th>ingredient</th>
    <th>calcium</th>
    <th>protein</th>
    <th>fibre</th>
    <th>cost(£) per pound </th>
</tr>
<tr>
    <td> limestone </td>   <td>0.380</td> <td>0.00</td> <td>0.00</td><td>0.0014</td>
</tr>
<tr>
    <td> corn </td>   <td>0.001</td> <td>0.09</td> <td>0.02</td><td>0.0463</td>
</tr>
<tr>
    <td> soybean meal </td>   <td>0.002</td> <td>0.5</td> <td>0.08</td><td>0.1250</td>
</tr>
</table>

Solution: Let $x_1,x_2,x_3$ be the amounts (in pounds of weight) of limestone corn and soybean meal used in producing the feed mix batch of 100 pounds. Then the linear programming model becomes
    <br>
    $x_0 = 0.0164x_1+0.0463x_2+0.1250x_3$
    <br><br>
    subject to the constraints $x_1+x_2+x3 = 100$ <br>
    $ 0.380x_1+0.001x_2+0.002x_3 \leq 0.012 \times 100$ <br>
    $ 0.380x_1+0.001x_2+0.002x_3 \geq 0.008 \times 100$ <br>
    $ 0.09x_2+0.5x_3 \leq 0.22 \times 100$ <br>
    $ 0.02x_2+0.08x_3 \leq 0.05 \times 100$ <br>
     $ x_1,x_2,x_3 \geq 0$ <br>

    

In [19]:
X0 = [0.0164,0.0463,0.1250]
A=[[1,1,1],[0.38,0.001,0.002],[-0.38,-0.001,-0.002],[0.0,0.09,0.5],[0.0,0.02,0.08]]
b=[100,1.2,-0.8,22,5]
res = linprog(X0, A_ub=A, b_ub=b,options={"disp": True, "maxiter":3})
print res

Optimization terminated successfully.
         Current function value: 0.034526    
         Iterations: 1
     fun: 0.034526315789473683
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([ 97.89473684,   0.4       ,  22.        ,   5.        ,   0.        ])
  status: 0
 success: True
       x: array([ 2.10526316,  0.        ,  0.        ])


In [103]:
print res

  status: 0
   slack: array([ 97.89473684,   0.4       ,  22.        ,   5.        ,   0.        ])
 success: True
     fun: 0.034526315789473683
       x: array([ 2.10526316,  0.        ,  0.        ])
 message: 'Optimization terminated successfully.'
     nit: 1


This result can be easily tested to be false, because the sum of the components of the x array do not add up to 100, as the first constraint is supposed to stipulate. This shall be fixed after subsequent examples.

<h2>Example 3</h2><br>
<p> In a factory there are 3 machines, M1, M2, and M3 used in making 2 products P1 and P2. One unit of P1 occupies M1 5 minutes, M2 3 minutes, and M3 4 minutes. The corresponding figures for one unit of P2 are: M1 1 minute, M2 4 minutes, M3 3 minutes. The net profit per unit of P1 produced is 30 pounds, and the net profit of each unit of p2 produced is 20 dollars (independent of whether the machines are used to full capacity. What production plan gives the most profit (assume that x1 units of P1 and x2 units of p2 are produced per hour</p> 

Ans: The problem is to maximize <br> $f=30x_1+20x_2$ <br>
subject to the constraints <br>
$5x_1+x_2 \leq 60$<br>
$3x_1+4x_2 \leq 60$<br>
$4x_1+3x_2 \leq 60$<br>
and $x_1,x_2 \geq 0$

This can also be expressed in matrix form <br>
$x_0=30x_1+20x_2$ <br><br>
$\mathbf {Ax}\leq\mathbf b$ <br><br>
$ \begin{pmatrix} 5&1\\3&4\\4&3\end{pmatrix}\begin{pmatrix} x_1\\x_2\end{pmatrix}\leq\begin{pmatrix}60\\ 60\\ 60 \end{pmatrix} $

<br>
And as the problem is to Maximize, the input function array must have negative coefficients [-30,-20] because the linprog function <i>minimizes</i>

<p style="color:red"> The Rows of the A matrix correspond to machine M1, M2 and M3 while the columns correspond to the products they produce </p> 

In [20]:
X0 = [-30,-20]
A=[[5,1],[3,4],[4,3]]
b=[60,60,60]
res = linprog(X0, A_ub=A, b_ub=b,options={"disp": True, "maxiter":3},callback = lambda *x,**kwargs: None)

Optimization terminated successfully.
         Current function value: -436.363636 
         Iterations: 2


In [21]:
print res

     fun: -436.36363636363637
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([ 0.        ,  5.45454545,  0.        ])
  status: 0
 success: True
       x: array([ 10.90909091,   5.45454545])


In [22]:
120.0/11.0

10.909090909090908

In [23]:
60./11.

5.454545454545454

The maximal profit is therefore 436.36 £/hour, the values of $x_1,x_2$ are [10.91, 5.45]

<h2>Example 4</h2><br>
A manufacturer procues a line of houshold products fabricated from sheet metal.To illustrate his production planning problem, consider that he makes 4 products and that his production system consists of 5 production centers: stamping, drilling, asssembly, finishing (painting and printing), and packaging. For a given month he must decide how much of each product to manufacture and to aid in this decision he has assembled the data shown in tables 1 and 2. Furthermore he knows that only 2000 square feet of the sheet metal used in products 2 and 4 will be available during the month. Product 2 requires 2.0 square feet per unit and product 4 uses 1.2 square feet per unit.

<br> formuate this as a linear programming model <br>
<h2> Table 1         Production data </h2>
<table>
<tr>
    <th>DEPARTMENT</th>
    <th>PRODUCT 1</th>
    <th>PRODUCT 2</th>
    <th>PRODUCT 3</th>
    <th>PRODUCT 4</th>
    <th>PRODUCTION HOURS AVAIlABLE</th>
</tr>
<tr>
    <td>Stamping</td>
    <td>0.03</td>
    <td>0.15</td>
    <td>0.05</td>
    <td>0.1</td>
    <td>400</td>
</tr>
<tr>
    <td>Drilling</td>
    <td>0.06</td>
    <td>0.12</td>
    <td>-</td>
    <td>0.1</td>
    <td>400</td>
</tr>
<tr>
    <td>Aseembly</td>
    <td>0.05</td>
    <td>0.1</td>
    <td>0.05</td>
    <td>0.12</td>
    <td>500</td>
</tr>
<tr>
    <td>Finishing</td>
    <td>0.04</td>
    <td>0.2</td>
    <td>0.03</td>
    <td>0.12</td>
    <td>450</td>
</tr>
<tr>
    <td>Packaging</td>
    <td>0.02</td>
    <td>0.06</td>
    <td>0.02</td>
    <td>0.05</td>
    <td>400</td>
</tr>
</table>
<br><br>
<h2> Table 2         Product data </h2>
<table>
<tr>
    <th>PRODUCT</th>
    <th>NET SELLING (PRICE/UNIT)</th>
    <th>VARIABLE (COST/UNIT)</th>
    <th>MINIMUM</th>
    <th>MAXIMUM</th>

</tr>
<tr>
    <td>1</td>
    <td>£10</td>
    <td>£6</td>
    <td>1000</td>
    <td>6000</td>

</tr>
<tr>
    <td>2</td>
    <td>25</td>
    <td>15</td>
    <td>-</td>
    <td>500</td>

</tr>
<tr>
    <td>3</td>
    <td>16</td>
    <td>11</td>
    <td>500</td>
    <td>3000</td>

</tr>
<tr>
    <td>4</td>
    <td>20</td>
    <td>14</td>
    <td>100</td>
    <td>1000</td>

</tr>

</table>



Solution: <br>
Let $x_i$ be the number of units of unit of product $i$ produced in the month and $Z$ be the total contribution to profit and overhead. The problem is to choose non-negative $x_1,x_2,x_3,x_4$ to maximize the fixed cost, that is the net selling price minus the varisble cost<br>
$Z=4x_1 +10x_2+5x_3+6x_4$ <br>
<br>and subject to the constraints on product hours for each department <br>

$0.03x_1+0.15x_2+0.05x_3+0.1x_4\leq400$                    Stamping<br>
$0.06x_1+0.12x_2\qquad\quad+0.1x_4\leq400$                    Drilling<<br>
$0.05x_1+0.1x_2+0.05x_3+0.12x_4\leq500$                    Assembly<br>
$0.04x_1+0.2x_2+0.03x_3+0.12x_4\leq450$                    Finishing<br>
$0.02x_1+0.06x_2+0.02x_3+0.05x_4\leq400$                   Packaging<br>

<br>
There is also a constraint on sheet metal availability. <br>
$2.0x_2+1.2x_4\leq 2000$
<br>
with the following bounds <br>
$1000\leq x_1\leq6000$<br>
$0\leq x_2\leq500$<br>
$500\leq x_3\leq3000$<br>
$100\leq x_4\leq1000$<br>

In [24]:
Z=[-4, -10, -5, -6, 0]
A=[[0.03, 0.15, 0.05, 0.1, 0.0],
   [0.06, 0.12, 0.0, 0.1, 0.0],
   [0.05, 0.1, 0.05, 0.12, 0.0],
   [0.04, 0.2, 0.03, 0.12, 0.0],
   [0.02, 0.06, 0.02, 0.05, 0.0],
   [0.0, 2.0, 0.0, 1.2, 0.0]]
b=[400, 400, 500, 450, 400, 2000]
x1_bounds = (1000, 6000)
x2_bounds = (0, 500)
x3_bounds = (500, 3000)
x4_bounds = (100, 1000)
x5_bounds = (0, 0)
res = linprog(Z, A_ub=A, b_ub=b, bounds=(x1_bounds, x2_bounds,x3_bounds,x4_bounds,x5_bounds),options={"disp": True})

Optimization terminated successfully.
         Current function value: -42600.000000
         Iterations: 9


In [25]:
print res

     fun: -42600.0
 message: 'Optimization terminated successfully.'
     nit: 9
   slack: array([   0.,    0.,   13.,   28.,  195.,  880.,  500.,    0.,    0.,
        900.,    0.,    0.,    0.,    0.])
  status: 0
 success: True
       x: array([ 5500.,   500.,  3000.,   100.,     0.])


The A matrix in this case was padded to contain extra zeros because the A matrix must have the same number of rows as columns of the input vector . Product $x_4$ is produced only because of the lower bound constraint $x_4 \geq 100$. This program uses the full capacity of stamping and drilling. The unused capacity in the remaining production centers is 13 hours in assembly, 28 hours in finishing, and 195 hours in packaging. Only 1120 square feet of sheet metal are used, leaving 880 unused,. The maximum profit is £42600.<br>
By examining values of the dual vriables, show how it would not be desirable to produce more of product 4 (at the expense of reducing production of some other product) unlessits unit profit were at least £1.78 higher, or £7.78. The marginal value of an additional hour of stamping capacity is £22.22 and of drilling capacity £55.56. If one could sell oe more unit of product 3 one could increase the profit by £3.89 (accounting for the fact tht one other product would have to be cut back) The ability to sell more of product 2 would yeild no profit, explain.

<br><br><h2>Revisiting example 2 </h2><br>
We can exress the equality $x_1+x_2+x_3=100$ as another row in the A matrix, by inserting two expressions <br>
$x_1+x_2+x_3\leq101$ <br>
$-x_1-x_2-x_3\leq-99$
<br><br>
and padding the A matrix with extra columns of zeros to make it square. This answer will only ever approximate the correct solution, so to be correct to several significant figures we can use numbers like 100.000001 and 99.999999

In [26]:
X0 = [0.0164,0.0463,0.1250,0.0,0.0,0.0]

A=[[ 1  ,1,     1      ,0,0,0],
   [-1  ,-1   ,-1      ,0,0,0],
   [0.38 ,0.001 ,0.002 ,0,0,0],
   [-0.38,-0.001,-0.002,0,0,0],
   [0.0  ,0.09  ,0.5   ,0,0,0],
   [0.0  ,0.02  ,0.08  ,0,0,0]]

b=[100.01,-100.00,1.2,-0.8,22.0,5.0]
res = linprog(X0, A_ub=A, b_ub=b,options={"disp": True})

Optimization terminated successfully.
         Current function value: 4.543219    
         Iterations: 3


In [27]:
print res

     fun: 4.5432189973614774
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([  1.00000000e-02,   0.00000000e+00,   1.32612137e+01,
         3.05804749e+00,   0.00000000e+00,   0.00000000e+00])
  status: 0
 success: True
       x: array([  2.90237467,  97.09762533,   0.        ,   0.        ,
         0.        ,   0.        ])


so apparently, we need to feed the animals only limestone and corn ... the sum of the components of the x array add up to 100 (or close enough within tolerable significant figures, give or take a grain or two)

<h3>Example 5</h3><br>
Assume there are three factories F1,F2 and F3 suupying goods to three warehouses W1,W2,and W3. The amounts available in each factory , the amounts needed in each warehouse and the costs of shipping from factory i to warehouse j are given in the table. Find the minimum cost of satisfying warehouse demands given that any factory may ship to any warehouse.<br><br>
<h2> Table </h2>
<table>
<tr>
<th></th>
<th>F1</th>
<th>F2</th>
<th>F3</th>
<th>Units Demanded</th>
</tr>
<tr>
    <td>W1</td>
    <td>£.90</td>
    <td>£1.00</td>
    <td>£1.00</td>
    <td>5</td>
</tr>
<tr>
    <td>W2</td>
    <td>£1.00</td>
    <td>£1.40</td>
    <td>£.80</td>
    <td>20</td>
</tr>
<tr>
    <td>W3</td>
    <td>£1.30</td>
    <td>£1.00</td>
    <td>£.80</td>
    <td>20</td>
</tr>
<tr>
    <td>Units Available</td>
    <td>20</td>
    <td>15</td>
    <td>10</td>
    <td>45</td>
</tr>
</table>

Solution: this problem is not easy to formulate in terms of the nodal incidence matrix methods typically used in scipy.optimize so I will import Pulp for this particular example.

In [33]:
from pulp import *

In [70]:
# Creates a list of all the supply nodes
Factories = [0, 1,2]

# Creates a list for the number of units of supply for each supply node
supply = [ 20,
         15,
         10]

# Creates a list of all demand nodes
Warehouses = [0, 1,2]

# Creates a list for the number of units of demand for each demand node
demand = [ 5,
           20,
           20
         ]


In [61]:
# creating the costs matrix such that costs [F][W] returns factory F to warehouse W
costs = [[0.9, 1.0, 1.3],
         [1.0, 1.4, 1.0],
         [1.0, 0.8, 0.8]
]
print costs[1][2]

1.0


In [73]:
# Creates the prob variable to contain the problem data
prob = LpProblem("Factory To Warehouse Trasnportation Problem2",LpMinimize)

In [74]:
# Creates a list of tuples containing all the possible routes for transport
Routes = [(f,w) for f in Factories for w in Warehouses]

In [75]:
# A dictionary called route_vars is created to contain the referenced variables (the routes)
route_vars = LpVariable.dicts("Route",(Factories,Warehouses),0,None,LpInteger)

In [84]:
print route_vars

{0: {0: Route_0_0, 1: Route_0_1, 2: Route_0_2}, 1: {0: Route_1_0, 1: Route_1_1, 2: Route_1_2}, 2: {0: Route_2_0, 1: Route_2_1, 2: Route_2_2}}


In [76]:
# The objective function is added to prob first
prob += lpSum([route_vars[f][w]*costs[f][w] for (f,w) in Routes]), "Sum of Transporting Costs"

In [77]:
print Routes

[(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]


In [78]:
# The supply maximum constraints are added to prob for each supply node (warehouse)
for f in Factories:
    prob += lpSum([route_vars[f][w] for w in Warehouses]) <= supply[f], "Sum of Products out of Factory %s"%f

# The demand minimum constraints are added to prob for each demand node (bar)
for w in Warehouses:
    prob += lpSum([route_vars[f][w] for f in Factories]) >= demand[w], "Sum of Products into Warehouse %s"%w

In [79]:
# The status of the solution is printed to the screen
print "Status:", LpStatus[prob.status]

Status: Not Solved


In [80]:
prob.solve()

1

In [81]:
print "Status:", LpStatus[prob.status]

Status: Optimal


In [82]:
for v in prob.variables():
    print v.name, "=", v.varValue

Route_0_0 = 5.0
Route_0_1 = 15.0
Route_0_2 = 0.0
Route_1_0 = 0.0
Route_1_1 = 0.0
Route_1_2 = 15.0
Route_2_0 = 0.0
Route_2_1 = 5.0
Route_2_2 = 5.0


In [83]:
print "Total Cost of Routes = ", value(prob.objective)

Total Cost of Routes =  42.5


As we can see the cost of the cheapest (optimal) route is £42.50 ... The table below shows the supply quantities <br>
<br>
<table>
    <tr>
        <td></td>
        <td>Factory 1</td>
        <td>Factory 2</td>
        <td>Factory 3</td>
    </tr>
    <tr>
        <td>Warehouse 1</td>
        <td>5</td>
        <td>-</td>
        <td>-</td>
    </tr>
    <tr>
        <td>Warehouse 2</td>
        <td>15</td>
        <td>-</td>
        <td>5</td>
    </tr>
    <tr>
        <td>Warehouse 3</td>
        <td>-</td>
        <td>15</td>
        <td>5</td>
    </tr>
</table>